pkg/manager/manager.go

```diff
@@ -131,56 +143,82 @@ func (r *LayerManager) ResolverMetaLayer(ctx context.Context, refspec reference.
 		target.Annotations[label.CRIImageRef] = refspec.String()
 		target.Annotations[label.CRILayerDigest] = target.Digest.String()
 
-		workdir := r.nydusFs.UpperPath(rawRef)
-		if _, err := os.Stat(workdir); os.IsNotExist(err) {
+		if _, ok = r.nydusMetaLayer.Load(snapshotID); ok {
+			log.G(ctx).Warnf("nydus duplicate mount meta layer ref is %s digest is %s", refspec.String(), target.Digest.String())
+			return &target, nil
+		}
+
+		workdir := r.nydusFs.UpperPath(snapshotID)
+		if _, err = os.Stat(workdir); os.IsNotExist(err) {
 			if err = os.MkdirAll(workdir, 0755); err != nil {
 				log.G(ctx).Errorf("mkdir nydus snapshot dir failed: %+v", err)
 				return nil, err
 			}
 		}
 
 		// Download nydus bootstrap layer to disk.
-		err = r.nydusFs.PrepareMetaLayer(ctx, storage.Snapshot{ID: rawRef}, target.Annotations)
+		err = r.nydusFs.PrepareMetaLayer(ctx, storage.Snapshot{ID: snapshotID}, target.Annotations)
 		if err != nil && !strings.Contains(err.Error(), "file exists") {
 			log.G(ctx).Errorf("download snapshot files failed: %+v", err)
 			return nil, err
 		}
 
+		nydusMsgChannel := make(chan nydusMessage)
+
 		go func() {
 			log.G(ctx).Debugf("nydus mount meta layer ref is %s digest is %s", refspec.String(), target.Digest.String())
-
-			if _, ok := r.nydusMetaLayer.Load(refspec.String()); ok {
-				log.G(ctx).Warnf("nydus duplicate mount meta layer ref is %s digest is %s", refspec.String(), target.Digest.String())
-				return
-			}
-
-			err = r.nydusFs.Mount(ctx, rawRef, target.Annotations)
+			err = r.nydusFs.Mount(ctx, snapshotID, target.Annotations)
 			if err != nil {
 				log.G(ctx).Errorf("nydus mount failed: %+v", err)
+				nydusMsgChannel <- nydusMessage{
+					Err:    err,
+					Status: MountFailed,
+				}
 				return
 			}
 
-			// Link nydusd mount dir to <mountpoint>/<ref>/<digest>/<diff>
-			targetPath := fmt.Sprintf("%s/store/%s/%s/diff", r.rootDir, rawRef, target.Digest.String())
-			mountPoint, err := r.nydusFs.MountPoint(rawRef)
-			if err == nil {
-				cmd := exec.Command("mount", "-o", "bind,ro", mountPoint, targetPath)
-				err = cmd.Run()
+			nydusMsgChannel <- nydusMessage{
+				Err:    nil,
+				Status: MountSuccess,
+			}
+		}()
+
+		event := <-nydusMsgChannel
+		if event.Err == nil {
```

Comment:
Just a code style question, why not check `event.Err ! = nil`, which would avoid a large indent of the code.